hera_sim/simulate.py

```diff
@@ -0,0 +1,335 @@
+            assert (
+                n_freq is not None
+            ), "if data_filename not given, n_freq must be given"
+            assert (
+                n_times is not None
+            ), "if data_filename not given, n_times must be given"
+            assert (
+                antennas is not None
+            ), "if data_filename not given, antennas must be given"
+            assert (
+                ant_pairs is not None
+            ), "if data_filename not given, ant_pairs must be given"
+
+            # Default values for mutable parameters
+            if pols is None:
+                pols = ["xx"]
+            if telescope_location is None:
+                telescope_location = io.HERA_LOCATION
+            if telescope_lat_lon_alt is None:
+                telescope_lat_lon_alt = io.HERA_LAT_LON_ALT
+
+            # Actually create it
+            self.data = io.empty_uvdata(
+                nfreq=n_freq,
+                ntimes=n_times,
+                ants=antennas,
+                antpairs=ant_pairs,
+                pols=pols,
+                time_per_integ=time_per_integ,
+                min_freq=min_freq,
+                channel_bw=channel_bw,
+                instrument=instrument,
+                telescope_location=telescope_location,
+                telescope_lat_lon_alt=telescope_lat_lon_alt,
+                object_name=object_name,
+                start_jd=start_jd,
+                vis_units=vis_units,
+            )
+
+        else:
+            self.data = self._read_data(self.data_filename, pols, **kwargs)
+
+    def _read_data(self, filename, polarizations, **kwargs):
+        uv = UVData()
+        uv.read(filename, polarizations=polarizations, read_data=True, **kwargs)
+        return uv
+
+    def write_data(self, filename, file_type=None, **kwargs):
+        """
+        Write current UVData object to file.
+
+        Args:
+            filename (str): filename to write to. Suffix of filename will determine the type of file to be written,
+                unless `file_type` is set.
+            file_type: (str): one of "miriad", "uvfits" or "uvh5" (i.e. any of the supported write methods of
+                :class:`pyuvdata.UVData`) which determines which write method to call.
+            **kwargs: keyword arguments sent directly to the write method chosen.
+        """
+        if file_type is None:
+            try:
+                file_type = filename.split(".")[-1]
+            except IndexError:
+                raise ValueError(
+                    "if no file_type is given, the filename must be suffixed with either .miriad, .uvfits or .uvh5"
+                )
+
+        if file_type not in ["miriad", "uvfits", "uvh5"]:
+            raise ValueError("file_type must be one of 'miriad', 'uvfits' or 'uvh5'")
+
+        getattr(self.data, "write_%s" % filename.split(".")[-1])(filename, **kwargs)
+
+    def _iterate_baselines(self, with_conj=True):
```

Comment:
This function iterates over all possible baselines based on `antenna_numbers`, which might include a lot of baselines that don't actually exist in the object for a number of reasons. I think we can solve two problems at once by making this something like
```
def _iterate_antpairpols(self):
  for app in self.data.get_antpairpols()
    blt_inds = self.data.antpair2ind(app)
    pol_ind = self.data.get_pols().index(app[2])
    yield app[0], app[1], app[2], blt_inds, pol_ind
```
Once we do that, in each `add_*` function we can use the relevant pol_ind in the `self.data.data_array[:, :, :, pol_ind]]` call, does that make sense?

Reply:
Yeah, I really was just taking a stab at it. This looks like a good addition. I've added it in 5e528d5a9505e6d5df19ae9fd57c424265a3d5c4